File: 2-Android With Java/MVC/Day5_Threads/app/src/main/java/com/example/day5_threads/favproducts/controller/FavActivity.java
package com.example.day5_threads.favproducts.controller;

import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.day5_threads.R;
import com.example.day5_threads.favproducts.view.FavAdapter;
import com.example.day5_threads.favproducts.view.OnFavoriteRemoveClickListener;
import com.example.day5_threads.model.DB.ProductDao;
import com.example.day5_threads.model.DB.ProductDataBase;
import com.example.day5_threads.model.Pojo.Product;
import com.google.android.material.snackbar.Snackbar;

import java.util.List;

public class FavActivity extends AppCompatActivity implements OnFavoriteRemoveClickListener {

    private RecyclerView recyclerView;
    private FavAdapter adapter;
    private ProductDao dao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fav_list);

        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        ProductDataBase db = ProductDataBase.gatInstance(this);
        dao = db.getProductDao();

        LiveData<List<Product>> productList = dao.getAllItems();
        productList.observe(this, new Observer<List<Product>>() {
            @Override
            public void onChanged(List<Product> products) {
                adapter = new FavAdapter(FavActivity.this, products, FavActivity.this);
                recyclerView.setAdapter(adapter);
            }
        });
    }

    @Override
    public void onFavRemoveProductClick(Product product)
    {

        new Thread(() -> {
            dao.deleteItem(product);
        }).start();
        adapter.notifyDataSetChanged();

        // Show Snackbar with undo action
        Snackbar snackbar = Snackbar.make(
                findViewById(android.R.id.content),
                "Item removed",
                Snackbar.LENGTH_LONG
        );

        snackbar.setAction("UNDO", v -> {
            // Undo the deletion
            new Thread(() -> {
                dao.insertItem(product);
            }).start();
            adapter.notifyDataSetChanged();
        });

        snackbar.show();
    }



}

File: 5-Architecture design patterns/MVVM/Products/app/src/main/java/com/example/products/features/favourite/view/FavActivity.kt
package com.example.products.features.favourite.view

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.day1.model.local.LocalDatabase
import com.example.day1.model.remote.IProductService
import com.example.products.R
import com.example.products.data.local.ProuctLocalDataSourceImpl
import com.example.products.data.remote.ProductRemoteDataSourceImpl
import com.example.products.data.remote.RetrofitHelper
import com.example.products.data.repo.ProductsRepoImpl
import com.example.products.databinding.FavListBinding
import com.example.products.databinding.ProductsListBinding
import com.example.products.features.allproducts.view.AllProductsAdaptor
import com.example.products.features.allproducts.viewmodel.AllProductViewModel
import com.example.products.features.allproducts.viewmodel.AllProductViewModelFactory
import com.example.products.features.favourite.viewmodel.FavViewModel
import com.example.products.features.favourite.viewmodel.FavViewModelFactory

class FavActivity : AppCompatActivity() {

    lateinit var viewModel: FavViewModel
    lateinit var vmFactory: FavViewModelFactory
    lateinit var favAdaptor: FavAdapter

    lateinit var binding : FavListBinding

    @SuppressLint("MissingPermission")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
//        setContentView(R.layout.fav_list)
        binding = FavListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        vmFactory= FavViewModelFactory(
            ProductsRepoImpl.getInstance(
                ProductRemoteDataSourceImpl(RetrofitHelper.retrofit.create(IProductService::class.java ) ), ProuctLocalDataSourceImpl(
                    LocalDatabase.getInstance(this).getProductDao())
            ))

        viewModel = ViewModelProvider(this@FavActivity , vmFactory).get(FavViewModel::class.java)
//        viewModel.getAllProducts()

        viewModel.getAllProducts()

        favAdaptor = FavAdapter { product ->
            viewModel.removeProduct(product)
            viewModel.getAllProducts()
        }

        binding.recyclerView.apply {
            adapter = favAdaptor
            layoutManager = LinearLayoutManager(this@FavActivity).apply {
                orientation= RecyclerView.VERTICAL
            }


            viewModel.onlineProducts.observe(this@FavActivity){ products ->
                favAdaptor.submitList(products)
            }

        }

    }
}